indexd/src/slabs.rs

```diff
@@ -162,32 +249,197 @@ pub struct SharedSlab {
+            length: val.length,
+        }
+    }
+}
+
+/// A SharedObject is an object that can be shared with others. It contains the encryption key
+/// needed to decrypt the data, as well as the slabs that make up the object.
+///
+/// It has no public fields to avoid corruption of the internal state.
 pub struct SharedObject {
-    pub key: String,
-    pub slabs: Vec<SharedSlab>,
-    #[serde_as(as = "Option<Base64>")]
-    pub meta: Option<Vec<u8>>,
+    encryption_key: EncryptionKey,
+    slabs: Vec<SharedSlab>,
+    metadata: Option<Vec<u8>>,
 }
 
 impl SharedObject {
+    pub fn new(
+        encryption_key: EncryptionKey,
+        slabs: Vec<SharedSlab>,
+        metadata: Option<Vec<u8>>,
+    ) -> Self {
+        SharedObject {
+            encryption_key,
+            slabs,
+            metadata,
+        }
+    }
+
+    pub fn slabs(&self) -> &Vec<SharedSlab> {
+        &self.slabs
+    }
+
+    pub fn metadata(&self) -> Vec<u8> {
+        self.metadata.clone().unwrap_or_default()
+    }
+
+    /// Computes the total size of the object by summing the lengths of its slabs.
     pub fn size(&self) -> u64 {
         self.slabs.iter().fold(0_u64, |v, s| v + s.length as u64)
     }
+
+    /// Returns a writer that decrypts data on-the-fly using the object's encryption key.
+    pub fn writer<W: AsyncWrite + Unpin>(&self, w: W, offset: usize) -> CipherWriter<W> {
+        CipherWriter::new(w, self.encryption_key.clone(), offset)
+    }
+
+    /// Converts the SharedObject into an Object that can be saved to an indexer.
+    /// The slabs must be pinned before the object can be saved.
+    pub fn object(&self) -> Object {
+        Object {
+            encryption_key: self.encryption_key.clone(),
+            slabs: self
+                .slabs
+                .iter()
+                .map(|s| SlabSlice {
+                    slab_id: s.id,
+                    offset: s.offset,
+                    length: s.length,
+                })
+                .collect(),
+            metadata: self.metadata.clone().unwrap_or_default(),
+            created_at: Utc::now(),
+            updated_at: Utc::now(),
+        }
+    }
 }
 
-/// UploadMeta represents an uploaded object and the metadata needed to
-/// retrieve it.
-#[derive(Debug, PartialEq)]
-pub struct UploadMeta {
-    pub encryption_key: EncryptionKey,
-    pub object: Object,
+impl From<SharedObject> for Object {
+    fn from(val: SharedObject) -> Self {
+        Object {
+            encryption_key: val.encryption_key,
+            slabs: val
+                .slabs
+                .iter()
+                .map(|s| SlabSlice {
+                    slab_id: s.id,
+                    offset: s.offset,
+                    length: s.length,
+                })
+                .collect(),
+            metadata: val.metadata.unwrap_or_default(),
+            created_at: Utc::now(),
+            updated_at: Utc::now(),
+        }
+    }
+}
+
+pub(crate) mod encryption {
```

Comment:
I'd move this to its own file. `slabs.rs` seems to not do its contents justice anymore.